File: platforms/VS/C_Tests/Bmp_Tests.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "utils.hpp"

#include "../../../include/datheader.h"
#include "wrappers.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace Bmp_Tests
{
	TEST_CLASS(Bmp_Tests)
	{
	public:
		TEST_METHOD(Given_a_faulty_bmp_When_created_Then_an_exception_is_raised) {
			std::vector<uint8_t> inputFile = utils::ReadFile("badrle.bmp");
			bmp_p bmp = Test_Bmp_createFromBmp(inputFile.data(), inputFile.size());
			Assert::IsNull(bmp, L"Expected null return value");
			Assert::AreEqual(1, Test_GetExceptionCode(), L"Incorrect error code");
			Assert::AreEqual("Compressed data contains an error", Test_GetExceptionString(), L"Incorrect error message");
		}

		TEST_METHOD(Given_a_faulty_imm_When_created_Then_an_exception_is_raised) {
			std::vector<uint8_t> pixelFile = utils::ReadFile("tooshort.bmp");
			std::vector<uint8_t> paletteFile = utils::ReadFile("pal8out.pam");
			bmp_p bmp = Test_Bmp_createFromImmAndPam(pixelFile.data(), pixelFile.size(), paletteFile.data(), paletteFile.size());
			Assert::IsNull(bmp, L"Expected null return value");
			Assert::AreEqual(1, Test_GetExceptionCode(), L"Incorrect error code");
			Assert::AreEqual("Image buffer size is too short for valid Imm", Test_GetExceptionString(), L"Incorrect error message");
			Test_Bmp_destroy(bmp);
		}

		TEST_METHOD(Given_a_sound_bmp_When_used_to_create_an_imm_Then_the_correct_buffers_are_returned) {
			std::vector<uint8_t> inputFile = utils::ReadFile("pal8out.bmp");
			std::vector<uint8_t> expectedPixelFile = utils::ReadFile("pal8out.imm");
			std::vector<uint8_t> expectedPaletteFile = utils::ReadFile("pal8out.pam");
			bmp_p bmp = Test_Bmp_createFromBmp(inputFile.data(), inputFile.size());
			try {
				std::vector<uint8_t> actualPixelFile(Test_Imm_immSize(bmp));
				uint8_t result1 = Test_Imm_immBuffer(bmp, actualPixelFile.data(), actualPixelFile.size());
				std::vector<uint8_t> actualPaletteFile(Test_Imm_pamSize(bmp));
				uint8_t result2 = Test_Imm_pamBuffer(bmp, actualPaletteFile.data(), actualPaletteFile.size());
				Assert::AreEqual((uint8_t)true, result1, L"Incorrect result from ImmBuffer");
				Assert::AreEqual((uint8_t)true, result2, L"Incorrect result from PamBuffer");
				Assert::AreEqual(expectedPixelFile.size(), actualPixelFile.size(), L"Incorrect size for Imm buffer");
				Assert::AreEqual(expectedPaletteFile.size(), actualPaletteFile.size(), L"Incorrect size for Pam buffer");
				Assert::IsTrue(expectedPixelFile == actualPixelFile, L"Imm buffers do not match");
				Assert::IsTrue(expectedPaletteFile == actualPaletteFile, L"Pam buffers do not match");
			}
			catch (...) {

			}
			Test_Bmp_destroy(bmp);
		}

		TEST_METHOD(Given_a_sound_imm_and_pam_When_used_to_create_a_bmp_Then_the_correct_buffers_are_returned) {
			std::vector<uint8_t> inputPixelFile = utils::ReadFile("pal8out.imm");
			std::vector<uint8_t> inputPaletteFile = utils::ReadFile("pal8out.pam");
			std::vector<uint8_t> expectedFile = utils::ReadFile("pal8qnt.bmp");
			bmp_p bmp = Test_Bmp_createFromImmAndPam(inputPixelFile.data(), inputPixelFile.size(), inputPaletteFile.data(), inputPaletteFile.size());
			try {
				std::vector<uint8_t> actualFile(Test_Imm_size(bmp));
				uint8_t result = Test_Imm_buffer(bmp, actualFile.data(), actualFile.size());
				Assert::AreEqual((uint8_t)true, result, L"Incorrect result returned from Buffer");
				Assert::AreEqual(expectedFile.size(), actualFile.size(), L"Incorrect size for buffer");
				Assert::IsTrue(expectedFile == actualFile, L"Buffers do not match");
			}
			catch (...) {

			}
			Test_Bmp_destroy(bmp);
		}
	};
}

File: platforms/VS/CSharp_Tests/Bmp_Tests.cs
//using Microsoft.VisualStudio.TestTools.UnitTesting;
//using carbon14.FuryUtils;
//using System;
//using System.Linq;

//namespace CSharp_Tests
//{
//    [TestClass]
//    public class Bmp_Tests
//    {
//        [TestMethod]
//        public void Given_an_invalid_bmp_file_When_a_bmp_is_created_Then_an_exception_is_raised()
//        {
//            try
//            {
//                Bmp bmp = new Bmp(Utils.ReadFile("badrle.bmp"));
//                Assert.Fail("Exception was not thrown");
//            }
//            catch (FuryException ex)
//            {
//                Assert.AreEqual(ErrorCodes.INVALID_FORMAT, ex.ErrorCode, "Incorrect error code was set");
//                Assert.AreEqual("Compressed data contains an error", ex.Message, "Incorrect error message was set");
//            }
//            catch (Exception ex)
//            {
//                Assert.Fail($"Incorrect exception: {ex.Message} was thrown");
//            }
//        }

//        [TestMethod]
//        public void Given_an_invalid_imm_file_When_a_bmp_is_created_Then_an_exception_is_raised()
//        {
//            try
//            {
//                Bmp bmp = new Bmp(Utils.ReadFile("tooshort.bmp"), Utils.ReadFile("pal8out.pam"));
//                Assert.Fail("Exception was not thrown");
//            }
//            catch (FuryException ex)
//            {
//                Assert.AreEqual(ErrorCodes.INVALID_FORMAT, ex.ErrorCode, "Incorrect error code was set");
//                Assert.AreEqual("Image buffer size is too short for valid Imm", ex.Message, "Incorrect error message was set");
//            }
//            catch (Exception ex)
//            {
//                Assert.Fail($"Incorrect exception: {ex.Message} was thrown");
//            }
//        }

//        [TestMethod]
//        public void Given_a_valid_bmp_file_When_pixel_and_palette_buffers_are_requested_Then_the_correct_buffers_are_returned()
//        {
//            Bmp bmp = new Bmp(Utils.ReadFile("pal8out.bmp"));
//            byte[] expectedPixelBuffer = Utils.ReadFile("pal8out.imm");
//            byte[] expectedPaletteBuffer = Utils.ReadFile("pal8out.pam");
//            byte[] actualPixelBuffer = bmp.ImmBuffer;
//            byte[] actualPaletteBuffer = bmp.PamBuffer;
//            Assert.IsTrue(actualPixelBuffer.SequenceEqual(expectedPixelBuffer), "Imm buffer is not correct");
//            Assert.IsTrue(actualPaletteBuffer.SequenceEqual(expectedPaletteBuffer), "Pam buffer is not correct");
//        }

//        [TestMethod]
//        public void Given_a_valid_bmp_file_When_bmp_buffer_is_requested_Then_the_correct_buffer_is_returned()
//        {
//            Bmp bmp = new Bmp(Utils.ReadFile("pal8out.imm"), Utils.ReadFile("pal8out.pam"));
//            byte[] expectedBuffer = Utils.ReadFile("pal8qnt.bmp");
//            byte[] actualBuffer = bmp.Buffer;
//            Assert.IsTrue(actualBuffer.SequenceEqual(expectedBuffer), "Buffer is not correct");
//        }
//    }
//}
